tests/functional/adapter/test_constraints.py

```diff
@@ -0,0 +1,71 @@
+import pytest
+
+class DuckDBColumnEqualSetup:
+    @pytest.fixture
+    def int_type(self):
+        return "NUMBER"
+
+    @pytest.fixture
+    def schema_int_type(self):
+        return "INT"
+        
+    @pytest.fixture
+    def data_types(self, schema_int_type, int_type, string_type):
+        # sql_column_value, schema_data_type, error_data_type
+        return [
+            # DuckDB's cursor doesn't seem to distinguish between:
+            #   INT and NUMERIC/DECIMAL -- both just return as 'NUMBER'
+            #   TIMESTAMP and TIMESTAMPTZ -- both just return as 'DATETIME'
+            #   [1,2,3] and ['a','b','c'] -- both just return as 'list'
```

Comment:
What would happen in these cases?

Context: dbt is running an "empty" version of the model query (`where false limit 0`), and extracting column names + data types from the cursor description, _before_ it actually materializes the model. If it detects a diff, it will raise an error like:

```
08:58:44  Compilation Error in model my_model (models/my_model.sql)
08:58:44    This model has an enforced contract that failed.
08:58:44    Please ensure the name, data_type, and number of columns in your contract match the columns in your model's definition.
08:58:44
08:58:44    | column_name | definition_type | contract_type | mismatch_reason    |
08:58:44    | ----------- | --------------- | ------------- | ------------------ |
08:58:44    | my_col      | NUMBER          | TEXT          | data type mismatch |
```

If no mismatch is detected, and it's a table, dbt will now materialize it via:
```sql
create table {model_name} (my_col int);
insert into table (select {model_sql});
```
The database's type system is flexible by design, and is willing to do a lot of creative coercion:
```
D create table my_tbl (id int);
D insert into my_tbl (select 2.1 as id);
D select * from my_tbl;
┌───────┐
│  id   │
│ int32 │
├───────┤
│     2 │
└───────┘
```

It will still catch the more egregious mismatches, including for structs and lists:
```sql
D create or replace table other_tbl (struct_col struct(num int));
D insert into other_tbl (select {'num': 'text'} as struct_col);
Error: Conversion Error: Could not convert string 'text' to INT32
D insert into other_tbl (select ['some', 'strings'] as list_of_int);
D create or replace table other_tbl (list_of_int int[]);
Error: Conversion Error: Could not convert string 'some' to INT32
```